sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala

```diff
@@ -193,14 +218,34 @@ private[sql] class JacksonGenerator(
    *
    * @param row The row to convert
    */
-  def write(row: InternalRow): Unit = writeObject(writeFields(row, schema, rootFieldWriters))
+  def write(row: InternalRow): Unit = dataType match {
+    case st: StructType =>
+      writeObject(writeFields(row, st, rootFieldWriters))
+    case _ => throw new UnsupportedOperationException(
+      s"this api is only used when `JacksonGenerator` is initialized with `StructType`")
```

Comment:
`JacksonGenerator` can only be used to write out a row when initialized with `StructType`.